### PR TITLE
Marker service initial implementation

### PR DESCRIPTION
**Pull request by narayaruna**
Process the incoming reports and store the
no data state of a cw forward expression in
DynamoDB

### CODE REVIEW

File: iep-lwc-fwding-admin/src/main/scala/com/netflix/iep/lwc/fwd/admin/Api.scala [hunk 1]
```diff
@@ -53,13 +61,32 @@ class Api(
       post {
         entity(as[JsonNode]) { json =>
           complete {
-            logger.info(s"Received report $json")
-            HttpResponse(StatusCodes.OK)
+            Json
+              .decode[List[Report]](json)
+              .foldLeft(Future[Unit] {}) { (future, report) =>
```
[the remark this answers is not in the document]
Comment (narayaruna) [reply iep-lwc-fwding-admin/src/main/scala/com/netflix/iep/lwc/fwd/admin/Api.scala]: Yes, I had that dilemma too.

File: iep-lwc-fwding-admin/src/main/scala/com/netflix/iep/lwc/fwd/admin/ExpressionDetailsDao.scala [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+}
+
+case class ExpressionDetails(
+  expressionId: ExpressionId,
+  lastReportTs: Long,
+  fwdMetricInfo: Option[FwdMetricInfo],
+  error: Option[Throwable],
+  noDataAgeMins: Int,
+  noScalingPolicyMins: Int,
```
[the remark this answers is not in the document]
Comment (narayaruna) [reply iep-lwc-fwding-admin/src/main/scala/com/netflix/iep/lwc/fwd/admin/ExpressionDetailsDao.scala]: `lastReportTs` is actually Report.timestamp and reports can be posted even when Report.fwdMetricInfo is None. The `noDataAgeMins` is basically `latest report timestamp - prev report timestamp when fwdMetricInfo is None` ([code](https://github.com/narayaruna/iep-apps/blob/master/iep-lwc-fwding-admin/src/main/scala/com/netflix/iep/lwc/fwd/admin/MarkerService.scala#L113)).

[the remark this answers is not in the document]
Comment (brharrington) [reply iep-lwc-fwding-admin/src/main/scala/com/netflix/iep/lwc/fwd/admin/ExpressionDetailsDao.scala]: My point though is it shouldn't store an age, but store a timestamp for the relevant events and then compute the age at the point it is being queried. Age is relative to the current time when queried and the current time isn't known until a query is done.

[the remark this answers is not in the document]
Comment (narayaruna) [reply iep-lwc-fwding-admin/src/main/scala/com/netflix/iep/lwc/fwd/admin/ExpressionDetailsDao.scala]: Right. The variable name is confusing, it actually now represents the time period between last good report and the latest report with no data. Instead I can actually keep a list of events. In this case one event to keep track of the timestamp of the last good report and determine the age at query time.

File: iep-lwc-fwding-admin/src/main/scala/com/netflix/iep/lwc/fwd/admin/ExpressionDetailsDao.scala [hunk 3]
```diff
@@ -32,52 +32,44 @@ trait ExpressionDetailsDao {
...
-        .withNumber(LastReportTs, exprDetails.lastReportTs)
-        .withNumber(NoDataAgeMins, exprDetails.noDataAgeMins)
-        .withNumber(NoScalingPolicyAgeMins, exprDetails.noScalingPolicyMins)
+        .withPrimaryKey(ExpressionIdAttr, Json.encode(exprDetails.expressionId))
+        .withNumber(Timestamp, exprDetails.timestamp)
+        .withMap(Events, exprDetails.events.asJava)
 
     exprDetails.fwdMetricInfo.map { m =>
-      item = item.withJSON(FwdMetricInfo, Json.encode(m))
+      item = item.withString(FwdMetricInfoAttr, Json.encode(m))
```
Comment (brharrington) [inline iep-lwc-fwding-admin/src/main/scala/com/netflix/iep/lwc/fwd/admin/ExpressionDetailsDao.scala]: Just curious, what was the reason for this change?

File: iep-lwc-fwding-admin/src/main/scala/com/netflix/iep/lwc/fwd/admin/MarkerService.scala [hunk 4]
```diff
@@ -105,24 +102,26 @@ object MarkerServiceImpl extends StrictLogging {
...
+  ): ExpressionDetails = {
+    import ExpressionDetails._
+
+    ExpressionDetails(
+      report.id,
+      report.timestamp,
+      report.metric,
+      report.error,
+      report.metric
+        .map(_ => Map(DataFoundEvent -> report.timestamp))
```
Comment (brharrington) [inline iep-lwc-fwding-admin/src/main/scala/com/netflix/iep/lwc/fwd/admin/MarkerService.scala]: Does this not need any of the data from the `FwdMetricInfo` object?

File: iep-lwc-fwding-admin/src/main/scala/com/netflix/iep/lwc/fwd/admin/ExpressionDetailsDao.scala (hunk 3 above)
Comment (narayaruna) [reply iep-lwc-fwding-admin/src/main/scala/com/netflix/iep/lwc/fwd/admin/ExpressionDetailsDao.scala]: Puts fail when a string value is empty. I am thinking to use a different type, like Map if it is necessary in the future to query by any field in this type.

File: iep-lwc-fwding-admin/src/main/scala/com/netflix/iep/lwc/fwd/admin/MarkerService.scala (hunk 4 above)
Comment (narayaruna) [reply iep-lwc-fwding-admin/src/main/scala/com/netflix/iep/lwc/fwd/admin/MarkerService.scala]: May be. I am expecting some changes in the events captured when I am using these to query expressions for removal.

**Pull request merged**